gradle/lucene-dev/lucene-dev-repo-composite.gradle

```diff
@@ -0,0 +1,159 @@
+//   3) An auto-wired 'lucene' subfolder, if present. To skip auto-wiring, pass
+//      a blank value in step 2: "-Plucene.dev.path=".
+
+
+// This script is applied in settings.gradle and later at build time: these two contexts
+// are distinctively different and have separate (and limited) APIs.
+def configuringSettings = (rootProject instanceof org.gradle.api.initialization.ProjectDescriptor)
+
+// Accessor for -P properties from settings or at build time.
+BiFunction<String, String, String> resolvePropertyValue = { propertyName, defValue ->
+  if (configuringSettings) {
+    return settings.startParameter.projectProperties.getOrDefault(propertyName, defValue)
+  } else {
+    return project.properties.getOrDefault(propertyName, defValue)
+  }
+}
+
+def PROP_FORCE_VERSION="lucene.dev.version"
+def PROP_FORCE_PATH="lucene.dev.path"
+def READ_ACCESS_PROPERTY="lucene-dev-path.dir"
+
+
+def forcedLuceneVersion = resolvePropertyValue(PROP_FORCE_VERSION, null)
+if (forcedLuceneVersion != null) {
+  if (!configuringSettings) {
+    logger.lifecycle("Lucene version forced by -P${PROP_FORCE_VERSION}=${forcedLuceneVersion}")
+
+    allprojects {
+      repositories {
+        mavenLocal()
+      }
+
+      tasks.withType(Test) {
+        def userHome = System.properties.get('user.home')
+        systemProperty READ_ACCESS_PROPERTY, file("${userHome}/.m2/repository/org/apache/lucene").absolutePath
+      }
+
+      configurations.all {
+        resolutionStrategy.eachDependency {
+          if (requested.group == "org.apache.lucene") {
+            useVersion(forcedLuceneVersion)
+            because("Lucene version forced manually by 'lucene.dev.version' property.")
+          }
+        }
+      }
+    }
+  }
+
+  // Step 1: end resolution
+  return
+}
+
+def luceneDevRepo = null
+def defaultLuceneDevRepo = file("${rootDir}/lucene")
+String propertyValue = resolvePropertyValue(PROP_FORCE_PATH, null)
+if (propertyValue != null) {
+  // Step 2.
+  if (propertyValue.isBlank()) {
+    if (defaultLuceneDevRepo.exists() && configuringSettings) {
+      logger.lifecycle("Local Lucene development repository has been detected but won't be used.")
+    }
+  } else {
+    luceneDevRepo = file("${propertyValue}").absoluteFile
+    if (!luceneDevRepo.exists()) {
+      throw new GradleException("Lucene repository does not exist at: -P${PROP_FORCE_PATH}=${luceneDevRepo}.")
+    }
+  }
+} else if (defaultLuceneDevRepo.exists()) {
+  // Step 3
+  luceneDevRepo = defaultLuceneDevRepo
+}
+
+
+if (luceneDevRepo != null) {
+  // Allow turning off this auto-wiring via -Dlucene.dev.repo=false (can't be a -P property because
+  // at settings evaluation time we don't have project properties yet).
+  if (configuringSettings) {
+    // Include Lucene repository as a composite and substitute module names.
+    includeBuild(luceneDevRepo) {
+      dependencySubstitution {
+        def substitutionMap = [
```

Comment:
Seems like a shame to have to explicitly enumerate all these here. Is there a way to programmatically generate this list so that it changes dynamically with future module additions/removals? I'm guessing not, but I think you'd be in a better place to judge.

In order to simulate what would happen currently if a new module were added to the project and this substitutionMap weren't updated, I commented out one line from the existing config, and verified that (unsurprisingly) that module was indeed not substituted. As long as the versions are reasonably compatible, the build still proceeded ok, just pulling in the dev repo local dependencies and fetching the prerelease5 dependency for the other. If there's not a way to generate the substitutionMap programmatically, perhaps there'd be a way to do a deferred check to cause the build to fail, to reduce the chances of insidious, hard-to-debug issues going forward?

Reply:
I've no idea if there is a way to do it programmatically. Maybe. But I bet it won't be elegant. ;) 

bq. perhaps there'd be a way to do a deferred check

I think you could try to do it by resolving all configurations and then looking at dependency versions... but this does count as ugly (see note above).

Reply:
For the sake of considering something concrete ... I won't say it _isn't_ ugly, but this is something along the lines of what I had in mind:
```diff
diff --git a/gradle/lucene-dev/lucene-dev-repo-composite.gradle b/gradle/lucene-dev/lucene-dev-repo-composite.gradle
index e78c40900c6..20a82db8a48 100644
--- a/gradle/lucene-dev/lucene-dev-repo-composite.gradle
+++ b/gradle/lucene-dev/lucene-dev-repo-composite.gradle
@@ -1,4 +1,5 @@
 import java.util.function.BiFunction
+import java.util.stream.Collectors
 
 /*
  * Licensed to the Apache Software Foundation (ASF) under one or more
@@ -153,6 +154,20 @@ if (luceneDevRepo != null) {
       tasks.withType(Test) {
         systemProperty READ_ACCESS_PROPERTY, luceneDevRepo.absolutePath
       }
+      tasks.all {
+        doLast {
+         // as a sanity check, get the actual, dependency-substitution-aware versions
+         // for all lucene modules, and verify they all match.
+          def luceneVersions = rootProject.configurations.getByName('unifiedClasspath')
+             .incoming.resolutionResult.allComponents.stream()
+             .map(ResolvedComponentResult::getModuleVersion)
+             .filter(mvi -> 'org.apache.lucene'.equals(mvi.getGroup()))
+             .map(mvi -> mvi.getVersion()).distinct().collect(Collectors.toList())
+         if (luceneVersions.size() != 1) {
+           throw new GradleException("found conflicting Lucene module versions: ${luceneVersions}")
+         }
+       }
+      }
     }
   }
 }

```

Reply:
The ugliness comes from the fact that you're scanning a configuration you don't depend on (and do it on each and every task)... My gut feeling is that this is just plain wrong. Also, this plugs into palantir's configuration - not the "official" gradle configurations from the java plugin. I'd really leave it out for now as it's not the core part of this issue... A nicer (more "correct") way to do it would be to create a dependency from all tasks that potentially consume java plugin's configurations (JavaCompile type) to a single per-project task that has an explicit dependency on the implementation classpath, then scan that... Let's leave it for another issue though, shall we?

Reply:
I think I know how to solve it better too... but let's file another issue and move this bit there so that we can figure out the best way to do it. For now - let's leave the hardcoded list. Lucene modules don't change that often.

Reply:
Yes, this all makes sense. All looks good to me in its current state!